### ISSUE TITLE
torch.symeig()  get NAN gradient when loss backpropagation

### ISSUE THREAD

**Issue by Rpersie** [pytorch/pytorch#23133]
## Reproduction BUG code
```python
import torch

A = torch.tensor([[1.0124e+08, 9.9517e+07, 5.3546e+07, 7.5420e+07, 1.0620e+08, -0.0000e+00,
         -0.0000e+00, -0.0000e+00, -0.0000e+00, -0.0000e+00],
        [9.9517e+07, 1.0104e+08, 5.5640e+07, 7.5404e+07, 1.0777e+08, -0.0000e+00,
         -0.0000e+00, -0.0000e+00, -0.0000e+00, -0.0000e+00],
        [5.3546e+07, 5.5640e+07, 4.2558e+07, 4.2080e+07, 6.3944e+07, -0.0000e+00,
         -0.0000e+00, -0.0000e+00, -0.0000e+00, -0.0000e+00],
        [7.5420e+07, 7.5404e+07, 4.2080e+07, 5.8183e+07, 8.2586e+07, -0.0000e+00,
         -0.0000e+00, -0.0000e+00, -0.0000e+00, -0.0000e+00],
        [1.0620e+08, 1.0777e+08, 6.3944e+07, 8.2586e+07, 1.1964e+08, -0.0000e+00,
         -0.0000e+00, -0.0000e+00, -0.0000e+00, -0.0000e+00],
        [0.0000e+00, 0.0000e+00, 0.0000e+00, 0.0000e+00, 0.0000e+00, 1.0124e+08,
         9.9517e+07, 5.3546e+07, 7.5420e+07, 1.0620e+08],
        [0.0000e+00, 0.0000e+00, 0.0000e+00, 0.0000e+00, 0.0000e+00, 9.9517e+07,
         1.0104e+08, 5.5640e+07, 7.5404e+07, 1.0777e+08],
        [0.0000e+00, 0.0000e+00, 0.0000e+00, 0.0000e+00, 0.0000e+00, 5.3546e+07,
         5.5640e+07, 4.2558e+07, 4.2080e+07, 6.3944e+07],
        [0.0000e+00, 0.0000e+00, 0.0000e+00, 0.0000e+00, 0.0000e+00, 7.5420e+07,
         7.5404e+07, 4.2080e+07, 5.8183e+07, 8.2586e+07],
        [0.0000e+00, 0.0000e+00, 0.0000e+00, 0.0000e+00, 0.0000e+00, 1.0620e+08,
         1.0777e+08, 6.3944e+07, 8.2586e+07, 1.1964e+08]],requires_grad=True)
eigval, eigvec = torch.symeig(A,eigenvectors=True)
maxid = torch.argmax(eigval)
bvector = eigvec[:,maxid]
loss = torch.sum(bvector)
loss.backward()
print (loss)
print (A.grad)
``` 

## Behavior
```python
loss :  tensor(2.1871, grad_fn=<SumBackward0>)
A.grad: 
tensor([[nan, nan, nan, nan, nan, nan, nan, nan, nan, nan],
        [0., nan, nan, nan, nan, nan, nan, nan, nan, nan],
        [0., 0., nan, nan, nan, nan, nan, nan, nan, nan],
        [0., 0., 0., nan, nan, nan, nan, nan, nan, nan],
        [0., 0., 0., 0., nan, nan, nan, nan, nan, nan],
        [0., 0., 0., 0., 0., nan, nan, nan, nan, nan],
        [0., 0., 0., 0., 0., 0., nan, nan, nan, nan],
        [0., 0., 0., 0., 0., 0., 0., nan, nan, nan],
        [0., 0., 0., 0., 0., 0., 0., 0., nan, nan],
        [0., 0., 0., 0., 0., 0., 0., 0., 0., nan]])
```
When I use the  ```autograd.detect_anomaly()``` to detect the NAN in the gradient. 
The error is  **Function 'SymeigBackward' returned nan values in its 0th output.**

## Expected behavior
A.grad is not none

## Versions
python:  3.5
pytorch: 1.0.1.post2

**Comment by vishwakftw** [comment]
The most likely cause of this issue could be that the eigenvalues are not distinct, in which case, the current method of compute backwards for symeig leads to infs. Subsequent multiplication by 0 could therefore lead to NaNs.

The formula we use can be found here: https://arxiv.org/pdf/1001.1654.pdf

**Comment by Rpersie** [comment]
So how can I solve this problem?  My matrix do have the same eigenvalues. Thank you !

**Comment by vishwakftw** [comment]
I’m not sure if there are any simple ways to do this, unfortunately.

**Comment by Rpersie** [comment]
Thank you! So it means that pytorch can not solve this problem now ?  This matrix can not be backpropagated in pytorch.

**Comment by vishwakftw** [comment]
Unfortunately, not at the moment. Even TensorFlow has the same behavior.

**Comment by Rpersie** [comment]
enen, I have an ugly idea like this. I just add an constant matrix to original  matrix , which makes it have different eigvalues but have same eigvectors.  So the grad backpropagation is ok now .  Do you think this way is fine? Thank you !
```python
import torch
A = torch.tensor([[1.0124e+08, 9.9517e+07, 5.3546e+07, 7.5420e+07, 1.0620e+08, -0.0000e+00,
         -0.0000e+00, -0.0000e+00, -0.0000e+00, -0.0000e+00],
        [9.9517e+07, 1.0104e+08, 5.5640e+07, 7.5404e+07, 1.0777e+08, -0.0000e+00,
         -0.0000e+00, -0.0000e+00, -0.0000e+00, -0.0000e+00],
        [5.3546e+07, 5.5640e+07, 4.2558e+07, 4.2080e+07, 6.3944e+07, -0.0000e+00,
         -0.0000e+00, -0.0000e+00, -0.0000e+00, -0.0000e+00],
        [7.5420e+07, 7.5404e+07, 4.2080e+07, 5.8183e+07, 8.2586e+07, -0.0000e+00,
         -0.0000e+00, -0.0000e+00, -0.0000e+00, -0.0000e+00],
        [1.0620e+08, 1.0777e+08, 6.3944e+07, 8.2586e+07, 1.1964e+08, -0.0000e+00,
         -0.0000e+00, -0.0000e+00, -0.0000e+00, -0.0000e+00],
        [0.0000e+00, 0.0000e+00, 0.0000e+00, 0.0000e+00, 0.0000e+00, 1.0124e+08,
         9.9517e+07, 5.3546e+07, 7.5420e+07, 1.0620e+08],
        [0.0000e+00, 0.0000e+00, 0.0000e+00, 0.0000e+00, 0.0000e+00, 9.9517e+07,
         1.0104e+08, 5.5640e+07, 7.5404e+07, 1.0777e+08],
        [0.0000e+00, 0.0000e+00, 0.0000e+00, 0.0000e+00, 0.0000e+00, 5.3546e+07,
         5.5640e+07, 4.2558e+07, 4.2080e+07, 6.3944e+07],
        [0.0000e+00, 0.0000e+00, 0.0000e+00, 0.0000e+00, 0.0000e+00, 7.5420e+07,
         7.5404e+07, 4.2080e+07, 5.8183e+07, 8.2586e+07],
        [0.0000e+00, 0.0000e+00, 0.0000e+00, 0.0000e+00, 0.0000e+00, 1.0620e+08,
         1.0777e+08, 6.3944e+07, 8.2586e+07, 1.1964e+08]],requires_grad=True)
A1 = A.detach()
eigvalng, eigvecng = torch.symeig(A1,eigenvectors=True)
eps_matrix = torch.diag(torch.randn(10)*1e8)
eps_matrix = eigvecng.mm(eps_matrix)
eps_matrix = eps_matrix.mm(torch.transpose(eigvecng,0,1))
eigval, eigvec = torch.symeig(A+eps_matrix,eigenvectors=True)
maxid = torch.argmax(eigval)
bvector = eigvec[:,maxid]
loss = torch.sum(bvector)
loss.backward()
print (loss)
print (A.grad)
````

**Comment by vishwakftw** [comment]
Perturbations are fine, but it depends on the application.

How ill-conditioned is your matrix? The perturbation seems to be too huge.

**Comment by Rpersie** [comment]
My eigvalues are too big as
```python
tensor([2.2283e+05, 2.2283e+05, 1.4133e+06, 1.4133e+06, 2.3873e+06, 2.3873e+06,
        1.1886e+07, 1.1886e+07, 4.0675e+08, 4.0675e+08])

```
So if I just add a small value, the biggest eigvalue can not be changed. So it also have the same eigvalues.
Thank you!  
enen, may be I can normalize the matrix before doing eigdecomposition.

**Comment by soumith** [comment]
closing the issue because it's not a bug

**Comment by YovelRom** [comment]
I'm reopening this issue, since it is indeed a bug. The implementation is mathematically mistaken, and should not return NaN. The correct formula is shown in [this](https://arxiv.org/pdf/1909.02414.pdf#page=10&zoom=100,144,529) paper, equation 13: ![image](https://user-images.githubusercontent.com/61155501/87254759-9657f100-c48d-11ea-91b6-6116d04cd291.png). 
You can see that for each symmetrical matrix which had been decomposed and a function f was activated on its eigenvalues, the resulting gradient includes a matrix L whose value in the i,j place is (with l being the i'th eigenvalue, and l' the j'th), is (f(l) - f(l'))/(l - l'). This is the exact gradient pytorch currently uses, except it is stated in a way in which it's easy to see why NaN's should not be returned when two eigenvalues are identical. You'll notice that this is the derivative approximation, and at the limit l' = l it becomes the derivative definition (and indeed, in the diagonal places where by definition l = l', it is written as f'(l)). 

I created custom function as a way to bypass this problem, but as I'm no expert in Autograd it requires the user to provide f' as well as f, and it is much less efficient. The gradient correction is pretty straightforward, as it only requires to find the NaN's and replace them with f'(l). I would be grateful if someone could correct this bug. Thanks in advance!

**Comment by AJMHPC** [comment]
> I’m not sure if there are any simple ways to do this, unfortunately.

It may be possible to get around the degenerate eigenvalue problem through the application of conditional Lorentzian broadening as described below. This should increase gradient stability, but would incur some error and so should be optional.

![IMG_0031](https://user-images.githubusercontent.com/48359623/89997012-d75d4400-dc8b-11ea-848f-e74292868bfe.jpeg)

Note:  ε  = 1E-12

Reference: http://arxiv.org/abs/1710.08717
